### PR TITLE
PLNSRVCE-1307: switch from noncaching client to controller/caching client for pipelineruns

### PR DESCRIPTION
**Pull request by gabemontero**
Per discussion in that Jira and email, converting from a
- non caching client, always hit the API server, polling mechanism for retrieving pipelineruns
to a
- cache client, watch based, non-polling, minimize load on the api servier mechanism for retrieving pipelineruns

Both @adambkaplan and I have a lot of experience from OCP with the drawbacks on non caching clients and the strict adherence to watch based retrieval whenever possible.

Quite possibly the abnormal restarting of the exporter in prod (I think seen by several folks, but reported to me by @sayan-biswas and @enarha ) could be addressed by this.

Changes we most likely should pursue, but I chose not to in this iteration
- leverage controller-runtime's prometheus integration
- possibly shift to constant metrics

On the possible shift in metric types, I have placed TODO comments as a reminder, as well as to at least initiate conversation here.  If you have bandwidth this week @adambkaplan please do minimally chime in on that.

I am choosing on refraining on those so as to
- isolate the positive impact of the switch to watches/informers from leveraging of controller-runtime integration
- minimize the impact @bnallapeta making progress on the grafana dashboard work in the short term

I also left the current logging mechanism alone, vs. logging mechanisms more typical of controller-runtime based offerings.  I'll leave any switches to @ramessesii2 and the logging work he has.

### CODE REVIEW

**Comment by gabemontero** [comment]
I have verified both with `go run main.go` as well as using `dev_setup.sh` in pipeline service to use images from this branch along with the changes from https://github.com/openshift-pipelines/pipeline-service/pull/636

File: collector/collector.go [hunk 1]
```diff
@@ -20,94 +20,44 @@ import (
...
+	client client.Client
+}
 
-func NewCollector(logger log.Logger) (*PipelineServiceCollector, error) {
+func NewCollector(logger log.Logger, client client.Client) (*PipelineServiceCollector, error) {
+	//TODO should this be converted to a Desc so we can use constant metrics
 	durationScheduled := prometheus.NewGaugeVec(prometheus.GaugeOpts{
 		Name: "pipelinerun_duration_scheduled_seconds",
 		Help: "Duration in seconds for a PipelineRun to be scheduled.",
+		//TODO the uid will get us a unique entry, but wouldn't adding namespace help usability with humans consuming this metric?
```
Comment (adambkaplan) [inline collector/collector.go:40]: How are we not exploding the metrics stack??? If we need data on how long an individual pipeline ran, we should get it from Results.

File: collector/controller.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	}); err != nil {
+		level.Error(logger).Log("msg", "timed out waiting for taskrun CRD to be created", err)
+		return nil, err
+	}
+
+	options.Scheme = runtime.NewScheme()
+	if err := k8sscheme.AddToScheme(options.Scheme); err != nil {
+		return nil, err
+	}
+	if err := pipelinev1beta1.AddToScheme(options.Scheme); err != nil {
```
Comment (adambkaplan) [inline collector/controller.go:44]: Please add a TODO here - we should upgrade to Tekton v1 now that the API is available.

File: collector/fetch.go [hunk 3]
```diff
@@ -2,51 +2,21 @@ package collector
...
-	return pipelineRuns, nil
+func (c *PipelineServiceCollector) getPipelineRuns() ([]v1beta1.PipelineRun, error) {
+	//NOTE:  there is no need to filter queries based on counts like with the old non-caching, polling client implementation;
+	// if over time we think we need to filter at all, it should be based on labels ...appstudio based labels for example:
+	//"appstudio.openshift.io/pac-provision"
+	//"appstudio.openshift.io/application"
+	//"appstudio.openshift.io/component"
+	prList := v1beta1.PipelineRunList{}
+	opts := &client.ListOptions{Namespace: metav1.NamespaceAll}
+	err := c.client.List(context.Background(), &prList, opts)
```
Comment (adambkaplan) [inline collector/fetch.go:20]: Perhaps a future TODO: can this deal with pagination? What happens when we have 10,000 active pipelines on the cluster?

File: collector/pipelinerun.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	contextTimeout = 300 * time.Second
+)
+
+func SetupPipelineRunReconcilerWithManager(mgr ctrl.Manager, logger log.Logger) error {
+	//TODO if/when we move off of the stand alone prometheus integration, we wil create metrics using the
```
Comment (adambkaplan) [inline collector/pipelinerun.go:23]: Can you please elaborate here on what you mean by "standalone Prometheus integration?" Offhand I don't understand what this refers to.

File: collector/pipelinerun.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+type ReconcilePipelineRun struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	logger        log.Logger
+}
+
+func (r *ReconcilePipelineRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```
Comment (adambkaplan) [inline collector/pipelinerun.go:41]: Why do we need an empty controller? Is there an easier way to get a client backed by a lister cache?

**Review by adambkaplan — COMMENTED** [review]
A good start on adding cached client support. I left a few notes on follow ups and the changes that are in scope.

I do want to follow up separately on what labels we are using to _aggregate_ data. I have grave concerns regarding the scalability of what we are building here.

File: collector/collector.go (hunk 1 above)
Comment (gabemontero) [reply collector/collector.go:40]: Yes my comment did not go that far, but in my head at least ;-), I am on the same page with you wrt the fact the labeling should change here, from both usability as well as scalability, hence why I was calling it out.

As I trid to alude to in this PR's description, I would not be suprised if the way the metrics are defined also contributes to the exporter instability we are seeing in prod etc.  Just switching to the caching client vs. non caching client will help prove that out.

File: collector/fetch.go (hunk 3 above)
Comment (gabemontero) [reply collector/fetch.go:20]: I think my note up top already gets into that.  Under the covers controller runtime is no different than knative in that they ultimately use shared informers, and if you look at the tekton reconcilers, I seen no use of pagination.  10,000 PRs should be OK with proper cpu/mem config.

File: collector/collector.go (hunk 1 above)
Comment (gabemontero) [reply collector/collector.go:40]: I'll update the TODO to make sure the scalability concern is clear.  And I'll go ahead and open the story off the current epic to get into this.

File: collector/pipelinerun.go (hunk 4 above)
Comment (gabemontero) [reply collector/pipelinerun.go:23]: Sure I'll update the comment.

Preview:
- the controller Options struct used to create the manager has a metric bind address
- the "sigs.k8s.io/controller-runtime/pkg/metrics" package has a `Registry.MustRegister(statsCollector)` method that sets up the prometheus stack and wires your collector's Describe and Collect metrics to be called

File: collector/pipelinerun.go (hunk 5 above)
Comment (gabemontero) [reply collector/pipelinerun.go:41]: Because I have another story after this one that is going to add the throttled by quota and throttled on  node resources constraints that I've demo'ed in the past.  That checks conditions from the Pod that tekton transfers to the PipelineRun.

Hence I'm not interested in sorting out the client backed by a PipelineRun lister cache only thing that will only be in place for a short while.  Having this here could also help with debug in between the 2 stories if the need arises.

I'll add a TODO comment explaining this will be filled in soon.

File: collector/fetch.go (hunk 3 above)
Comment (gabemontero) [reply collector/fetch.go:20]: Also look at the RHTAP components like build service ... again no pagination, just label based selectors.

If need be we'll restrict on those lines .... much simpler than pagination.

File: collector/controller.go (hunk 2 above)
Comment (gabemontero) [reply collector/controller.go:44]: Sure I can do that, though I would think converting from v1beta1 to v1 is going to be a product wide Jira based thing.

File: collector/pipelinerun.go (hunk 5 above)
Comment (gabemontero) [reply collector/pipelinerun.go:41]: Forgot - I can do the new/upcoming stats with the prometheus/collect flow (vs. counter.Inc() etc.).

And if the godoc is accurate, `ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})` may be sufficient for creating the watch/lister cache, where bypassing the call to `Complete(reconciler)` means we will not get called on events.  I'll see if I can cobble together an experiment tomorrow to prove that.

**Review by bnallapeta — DISMISSED** [review]
LGTM barring the comments others have already mentioned.

**Comment by gabemontero** [comment]
> Question @gabemontero : Will it be an overhead (and may be not needed) to add logs whenever Metrics controller has a new PipelineRun in its radar? 

It should not be default logging level @ramessesii2 

Depending on which logging framework you'll move us to, it is OK to have that at a "debug" level or a "higher" klog level.

>I understand the purpose of metrics-exporter is to have those additional metrics in the Graffana and so it's okay not to have those logs?

Correct, taking into account my clarification above

File: collector/pipelinerun.go (hunk 5 above)
Comment (gabemontero) [reply collector/pipelinerun.go:41]: Had some fun with the debugger today @adambkaplan 

Using the code as is, I set break points in 
- k8s.io/client-go/tools/cache/shared_informer.go:NewSharedIndexInformer
- k8s.io/client-go/tools/cache/reflector.go:ListAndWatch

And got hits for both when bringing up the exporter.

Then, when I changed 

`ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{}) .Complete(reconciler)` 

to 

`ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})`

I got no hit.  But when I followed that with 

```
ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})
prList := v1beta1.PipelineRunList{}
opts := &client.ListOptions{Namespace: metav1.NamespaceAll}
mgr.GetClient().List(context.Background(), &prList, opts)
```

and restarted, I got hits on those break points again.

In hindsight, I now realize this lines up with controller-runtime's non-caching client options (`options.ClientDisableCacheFor`) where if you don't set that, controller-runtime's use of dynamic client includes setting up a caching client by default.

Bottom line,  I'll lose the empty reconciler :-)

Comment (gabemontero) [reply collector/pipelinerun.go:41]: actually I do not need `ctrl.NewControllerManagedBy(mgr).For(&v1beta1.PipelineRun{})` as well

**Comment by gabemontero** [comment]
updates from comments pushed @adambkaplan  @ramessesii2

File: collector/collector.go (hunk 1 above)
Comment (gabemontero) [reply collector/collector.go]: story https://issues.redhat.com/browse/PLNSRVCE-1310 opened

**Review by ramessesii2 — APPROVED** [review]

**Comment by adambkaplan** [comment]
/hold

Marking this on hold due to summit code freeze.

**Comment by gabemontero** [comment]
with the exporter scaled down in staging / prod, can we go ahead and unhold / complete vetting / merge so we can continue development of dashboards and use of this for metrics in the pipeline-service repo @adambkaplan ?

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by adambkaplan** [comment]
Merging so we can make further iterations with the cached client.
